Simplify BYOS section (#278)
Microsoft/api-guidelines: azure/Guidelines.md

:white_check_mark: **DO** look for **both** HEADERS in client code, preferring the `Operation-Location` version.

### Bring your own Storage (BYOS)

> Many services need to store and retrieve data files. For this scenario, teh service should not implement its own 

Reviewer: ```suggestion
Many services need to store and retrieve data files. For this scenario, the service should not implement its own 
```

> Many services need to store and retrieve data files. For this scenario, teh service should not implement its own 

Reviewer: My strong opinion is that BYOS should be an option not a requirement. Forcing the user to create and provide a storage account with the required permissions etc. drastically raises the bar for starting to use a simple service. This is also what service teams find when they talk to customers. Both service managed storage and BYOS are valid scenarios.

Reviewer 2: I read this as warning against adding your own API façade on top of Blobs like we see so many teams do.  If they want to go through the work for Managed Storage, that's great and will improve the getting started experience for their customers.  But if they try to add their own `UploadFile` API for files larger than around 256MB, they're probably going to do a much worse job than the Storage team's end to end developer experience and should just accept a SAS URI to a blob.

Author: Yes, I was trying to say what Ted is saying (don't write your own API on top of Azure storage).
I'm OK if teams want to use managed storage too. We can add a sentence to this effect in the intro but then the team is writing a storage API. We can discuss but I'm OK with all of this.

> storage mechanism and should instead leverage the existing Azure Storage service. When doing this, the customer 

Author: ```suggestion
storage APIs and should instead leverage the existing Azure Storage service. When doing this, the customer 
```

> "owns" the storage account and just tells your service to use it. Colloquially, we call this <i>Bring Your Own Storage</i> as the customer is bringing their storage account to another service. BYOS provides significant benefits to service implementors: security, performance, uptime, etc. And, of course, most Azure customers are already with the Azure Storage service. 

Reviewer 2: ```suggestion
"owns" the storage account and just tells your service to use it. Colloquially, we call this <i>Bring Your Own Storage</i> as the customer is bringing their storage account to another service. BYOS provides significant benefits to service implementors: security, performance, uptime, etc. And, of course, most Azure customers are already familiar with the Azure Storage service. 
```

Reviewer 3: +1 on Ted's suggestions.

While Azure Managed Storage may be easier to get started with, as your service evolves and matures, BYOS provides the most flexibility and implementation choices. Further, when designing your APIs, be cognizant of expressing storage concepts and how clients will access your data. For example, if you are working with blobs, then you should not expose the concept of folders.

:white_check_mark: **DO** use the Bring Your Own Storage pattern

> :white_check_mark: **DO** use a blob prefix for logical folders

Reviewer: What does this mean?

Author: I'm afraid that teams don't really understand how storage works. Specifically, that there are containers but no subcontainers and that there aren't sub-directories either. There are just blob names which have delimiters in them (like / or \) and the customer defines the delimiter. So, I agree that I have not said this well.  How about:
**DO** use a URL indicating a path to a root directory.

This doesn't explain anything but I did add a link to the List Blobs API docs in one bullet so maybe engineers will ready that to understand how it really works.

Reviewer 2: Is it worth adding "instead of APIs called `directory`, `folder`, or `path`" to make that clearer?

Author: ```suggestion
:white_check_mark: **DO** use a blob prefix for a logical folder (avoid terms such as ```directory```, ```folder```, or ```path```).
```

:no_entry: **DO NOT** require a fresh container per operation

:white_check_mark: **DO** use managed identity and Role Based Access Control ([RBAC](https://docs.microsoft.com/en-us/azure/role-based-access-control/overview)) as the mechanism allowing customers to grant permission to their Storage account to your service.

:white_check_mark: **DO** Add RBAC roles for every service operation that requires accessing Storage scoped to the exact permissions.

:white_check_mark: **DO** Ensure that RBAC roles are backward compatible, and specifically, do not take away permissions from a role that would break the operation of the service. Any change of RBAC roles that results in a change of the service behavior is considered a breaking change.

:heavy_check_mark: **YOU MAY** use Shared Access Signatures [SAS](https://docs.microsoft.com/en-us/azure/storage/common/storage-sas-overview) to provide access to Storage objects.

:ballot_box_with_check: **YOU SHOULD** if using HTTP (not HTTPS) document to users that all information is sent over the wire in clear text.

> :white_check_mark: **DO** return an HTTP status code representing the result of your service operation's behavior. If the operation uses Storage internally, this an implementation detail from the client's perspective of the operation but may happen due to a missing Storage object or insufficient permissions. If the internal Storage operation fails, include the Storage error information in the 'inner-error' section of your operation's response body to help the client know that they may have configured their Storage improperly. 

Reviewer 4: I feel like we might want to expend on this a bit in all manners regarding permissions, which is probably the No. 1 issue customers will encounter. Specifically, errors regarding setting service principles, CMK setup and key rotation, etc. Having a consistent experience here across service will be great.

Author: We think what you're suggesting is already covered by the text. If you think more should be here, can you please provide the actual text you'd like to see? Thanks.

:white_check_mark: **DO** allow the customer to specify a URL path to a single Storage object if your service requires access to a single file.

:heavy_check_mark: **YOU MAY** allow the customer to provide a [last-modified](https://datatracker.ietf.org/doc/html/rfc7232#section-2.2) timestamp (in RFC1123 format) for read-only files. This allows the client to specify exactly which version of the files your service should use. When reading a file, your service passes this timestamp to Azure Storage using the [if-unmodified-since](https://datatracker.ietf.org/doc/html/rfc7232#section-3.4) request header. If the Storage operation fails with 412, the Storage object was modified and your service operation should return an appropriate 4xx status code and return the Storage error in your operation's 'inner-error' (see guideline above).

> :white_check_mark: **DO** allow the customer to specify a URL path to a logical directory (via prefix and delimiter) if your service requires access to multiple files (within this directory). For more information, see [List Blobs API](https://docs.microsoft.com/en-us/rest/api/storageservices/list-blobs)

Reviewer 4: Should we expand about file/path selectors? like file prefixes/suffixes (e.g. `*.pdf`).

Reviewer 3: @assafi , would you expand on your scenario a bit? I'm not sure I'm fully understanding the issue/question you are raising.

Reviewer 4: In some cases APIs will point to a folder but will only want/allow to process certain files. For example, [Document Translator API](https://docs.microsoft.com/en-us/azure/cognitive-services/translator/document-translation/reference/start-translation#inputs) let's users filter which files they want to translate.

Author: ```suggestion
:white_check_mark: **DO** allow the customer to specify a URL path to a logical folder (via prefix and delimiter) if your service requires access to multiple files (within this folder). For more information, see [List Blobs API](https://docs.microsoft.com/en-us/rest/api/storageservices/list-blobs)

:heavy_check_mark: **YOU MAY** offer an ```extensions``` field representing an array of strings indicating file extensions of desired blobs within the logical folder. 
```

A common pattern when working with multiple files is for your service to receive requests that contain the location(s) of files to process ("input") and a location(s) to place any files that result from processing ("output"). Note: the terms "input" and "output" are just examples; use terms more appropriate to your service's domain.

For example, a service's request body to configure BYOS may look like this:

```json
{
  "input":{
    "location": "https://mycompany.blob.core.windows.net/documents/english/?<sas token>",

>     "delimiter": "/",

Author: ```suggestion
    "delimiter": "/",
    "extensions" : [ ".bmp", ".jpg", ".tif", ".png" ],
```

    "lastModified": "Wed, 21 Oct 2015 07:28:00 GMT"
  },
  "output":{
    "location": "https://mycompany.blob.core.windows.net/documents/spanish/?<sas token>",
    "delimiter":"/"
  }
}
```

Depending on the requirements of the service, there can be any number of "input" and "output" sections, including none. 

:white_check_mark: **DO** include a JSON object that has string values for "location" and "delimiter". For "location", the customer must pass a URL to a blob prefix which represents a directory. For "delimiter", the customer must specify the delimiter character they desire to use in the location URL; typically "/" or "\". 

:heavy_check_mark: **YOU MAY** support the "lastModified" field for input directories (see guideline above).

:white_check_mark: **DO** support a "location" URL with a container-scoped SAS that has a minimum of `listing` and `read` permissions for input directories.

> :white_check_mark: **DO** support a "location" URL with a container-scoped SAS that has a minimum of `write` permissions for output directories.

Reviewer 2: What are your thoughts on [`write` vs. `create`](https://docs.microsoft.com/en-us/rest/api/storageservices/create-service-sas#permissions-for-a-directory-container-or-blob) here?  Is `create` a little safer if you don't want to trash anything already in `output.location`?  Or should we recommend `write` because it'll always get the job done?

Author: We agreed to leave this as is

:white_check_mark: **DO** look for **both** HEADERS in client code, preferring the `Operation-Location` version.

### Bring your own Storage (BYOS)
Many services need to store and retrieve data files. For this scenario, the service should not implement its own 

> storage APIs and should instead leverage the existing Azure Storage service. When doing this, the customer 

Reviewer 5: Should we make “Azure Storage service” a link to the service documentation?

Author: ```suggestion
storage APIs and should instead leverage the existing [Azure Storage service](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction). When doing this, the customer 
```

"owns" the storage account and just tells your service to use it. Colloquially, we call this <i>Bring Your Own Storage</i> as the customer is bringing their storage account to another service. BYOS provides significant benefits to service implementors: security, performance, uptime, etc. And, of course, most Azure customers are already familiar with the Azure Storage service. 

> While Azure Managed Storage may be easier to get started with, as your service evolves and matures, BYOS provides the most flexibility and implementation choices. Further, when designing your APIs, be cognizant of expressing storage concepts and how clients will access your data. For example, if you are working with blobs, then you should not expose the concept of folders.

Reviewer 5: Is “Azure Managed Storage” different from the “Azure Storage service”? If so, will that be obvious to readers?  Do we need to clarify that?

:no_entry: **DO NOT** require a fresh container per operation

:white_check_mark: **DO** use managed identity and Role Based Access Control ([RBAC](https://docs.microsoft.com/en-us/azure/role-based-access-control/overview)) as the mechanism allowing customers to grant permission to their Storage account to your service.

:white_check_mark: **DO** Add RBAC roles for every service operation that requires accessing Storage scoped to the exact permissions.

:white_check_mark: **DO** Ensure that RBAC roles are backward compatible, and specifically, do not take away permissions from a role that would break the operation of the service. Any change of RBAC roles that results in a change of the service behavior is considered a breaking change.

:heavy_check_mark: **YOU MAY** use Shared Access Signatures [SAS](https://docs.microsoft.com/en-us/azure/storage/common/storage-sas-overview) to provide access to Storage objects.

:ballot_box_with_check: **YOU SHOULD** if using HTTP (not HTTPS) document to users that all information is sent over the wire in clear text.

> :white_check_mark: **DO** return an HTTP status code representing the result of your service operation's behavior. If the operation uses Storage internally, this an implementation detail from the client's perspective of the operation but may happen due to a missing Storage object or insufficient permissions. If the internal Storage operation fails, include the Storage error information in the 'inner-error' section of your operation's response body to help the client know that they may have configured their Storage improperly. 

Reviewer 5: I think we can make this a little more concise:
```suggestion
:white_check_mark: **DO** include the Storage error information in the `inner-error` section of an error response if the error was the result of an an internal Storage operation failure.  This may help the client determine the underlying cause of the error, e.g a missing Storage object or insufficient permissions.
```

Author: ```suggestion
:white_check_mark: **DO** return an HTTP status code representing the result of your service operation's behavior. 

:white_check_mark: **DO** include the Storage error information in the 'inner-error' section of an error response if the error was the result of an internal Storage operation failure. This helps the client determine the underlying cause of the error, e.g.: a missing storage object or insufficient permissions. 
```
